pydantic_ai_slim/pydantic_ai/mcp.py

```diff
@@ -255,11 +255,79 @@ async def direct_call_tool(
             # See https://github.com/modelcontextprotocol/python-sdk#structured-output
             if isinstance(structured, dict) and len(structured) == 1 and 'result' in structured:
-                return structured['result']
-            return structured
+                return (
+                    messages.ToolReturn(return_value=structured['result'], metadata=result.meta)
+                    if result.meta
+                    else structured['result']
+                )
+            return messages.ToolReturn(return_value=structured, metadata=result.meta) if result.meta else structured
 
-        mapped = [await self._map_tool_result_part(part) for part in result.content]
-        return mapped[0] if len(mapped) == 1 else mapped
+        mapped_part_metadata_tuple_list = [await self._map_tool_result_part(part) for part in result.content]
+        if (
+            all(mapped_part_metadata_tuple[1] is None for mapped_part_metadata_tuple in mapped_part_metadata_tuple_list)
+            and result.meta is None
+        ):
+            # There is no metadata in the tool result or its parts, return just the mapped values
+            return (
+                mapped_part_metadata_tuple_list[0][0]
+                if len(mapped_part_metadata_tuple_list[0]) == 1
+                else [mapped_part_metadata_tuple[0] for mapped_part_metadata_tuple in mapped_part_metadata_tuple_list]
+            )
+        elif (
+            all(mapped_part_metadata_tuple[1] is None for mapped_part_metadata_tuple in mapped_part_metadata_tuple_list)
+            and result.meta is not None
+        ):
+            # There is no metadata in the tool result parts, but there is metadata in the tool result
+            return messages.ToolReturn(
+                return_value=(
+                    mapped_part_metadata_tuple_list[0][0]
+                    if len(mapped_part_metadata_tuple_list[0]) == 1
+                    else [
+                        mapped_part_metadata_tuple[0] for mapped_part_metadata_tuple in mapped_part_metadata_tuple_list
+                    ]
+                ),
+                metadata=result.meta,
+            )
+        else:
+            # There is metadata in the tool result parts and there may be a metadata in the tool result, return a ToolReturn object
+            return_values: list[Any] = []
+            user_contents: list[Any] = []
+            return_metadata: dict[str, Any] = {}
+            for idx, (mapped_part, part_metadata) in enumerate(mapped_part_metadata_tuple_list):
+                if part_metadata is not None:
+                    # Merge the metadata dictionaries, with part metadata taking precedence
+                    if return_metadata.get('content', None) is None:
+                        # Create an empty list if it doesn't exist yet
+                        return_metadata['content'] = list[dict[str, Any]]()
```

Comment:
We can likely use `return_metadata.setdefault('content', [])` just once before the `for`